File: frontend/packages/api-client/valid_vessels.ts
import { Axios } from 'axios';
import format from 'string-format';

import {
  ValidVesselCreationRequestInput,
  ValidVessel,
  QueryFilter,
  ValidVesselUpdateRequestInput,
  QueryFilteredResult,
  APIResponse,
} from '@dinnerdonebetter/models';

import { backendRoutes } from './routes';

export async function createValidVessel(client: Axios, input: ValidVesselCreationRequestInput): Promise<ValidVessel> {
  return new Promise(async function (resolve, reject) {
    const response = await client.post<APIResponse<ValidVessel>>(backendRoutes.VALID_VESSELS, input);

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function getValidVessel(client: Axios, validVesselID: string): Promise<ValidVessel> {
  return new Promise(async function (resolve, reject) {
    const response = await client.get<APIResponse<ValidVessel>>(format(backendRoutes.VALID_VESSEL, validVesselID));

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function getValidVessels(
  client: Axios,
  filter: QueryFilter = QueryFilter.Default(),
): Promise<QueryFilteredResult<ValidVessel>> {
  return new Promise(async function (resolve, reject) {
    const response = await client.get<APIResponse<ValidVessel[]>>(backendRoutes.VALID_VESSELS, {
      params: filter.asRecord(),
    });

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    const result = new QueryFilteredResult<ValidVessel>({
      data: response.data.data,
      totalCount: response.data.pagination?.totalCount,
      page: response.data.pagination?.page,
      filteredCount: response.data.pagination?.filteredCount,
    });

    resolve(result);
  });
}

export async function updateValidVessel(
  client: Axios,
  validVesselID: string,
  input: ValidVesselUpdateRequestInput,
): Promise<ValidVessel> {
  return new Promise(async function (resolve, reject) {
    const response = await client.put<APIResponse<ValidVessel>>(
      format(backendRoutes.VALID_VESSEL, validVesselID),
      input,
    );

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function deleteValidVessel(client: Axios, validVesselID: string): Promise<ValidVessel> {
  return new Promise(async function (resolve, reject) {
    const response = await client.delete<APIResponse<ValidVessel>>(format(backendRoutes.VALID_VESSEL, validVesselID));

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function searchForValidVessels(client: Axios, query: string): Promise<ValidVessel[]> {
  return new Promise(async function (resolve, reject) {
    const searchURL = `${backendRoutes.VALID_VESSELS_SEARCH}?q=${encodeURIComponent(query)}`;
    const response = await client.get<APIResponse<ValidVessel[]>>(searchURL);

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

File: backend/pkg/apiclient/valid_vessels.go
package apiclient

import (
	"context"

	"github.com/dinnerdonebetter/backend/internal/observability"
	"github.com/dinnerdonebetter/backend/internal/observability/keys"
	"github.com/dinnerdonebetter/backend/internal/observability/tracing"
	"github.com/dinnerdonebetter/backend/pkg/apiclient/generated"
	"github.com/dinnerdonebetter/backend/pkg/types"
)

// var (\w+) types\.APIResponse\[\*types\.(\w+)\]\n\	if err \= c\.fetchAndUnmarshal\(ctx\, req\, \&(\w+))\)\; err \!\= nil \{

// GetValidVessel gets a valid vessel.
func (c *Client) GetValidVessel(ctx context.Context, validVesselID string) (*types.ValidVessel, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if validVesselID == "" {
		return nil, ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.ValidVesselIDKey, validVesselID)
	tracing.AttachToSpan(span, keys.ValidVesselIDKey, validVesselID)

	res, err := c.authedGeneratedClient.GetValidVessel(ctx, validVesselID)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "retrieving valid vessel")
	}
	defer c.closeResponseBody(ctx, res)

	var apiResponse types.APIResponse[*types.ValidVessel]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "loading valid vessel response")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	return apiResponse.Data, nil
}

// GetRandomValidVessel gets a valid vessel.
func (c *Client) GetRandomValidVessel(ctx context.Context) (*types.ValidVessel, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	res, err := c.authedGeneratedClient.GetRandomValidVessel(ctx)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "retrieving valid vessel")
	}
	defer c.closeResponseBody(ctx, res)

	var apiResponse types.APIResponse[*types.ValidVessel]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "retrieving valid vessel")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	return apiResponse.Data, nil
}

// SearchValidVessels searches through a list of valid vessels.
// TODO: add queryFilter param here.
func (c *Client) SearchValidVessels(ctx context.Context, query string, limit uint8) ([]*types.ValidVessel, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if query == "" {
		return nil, ErrEmptyQueryProvided
	}

	if limit == 0 {
		limit = types.DefaultQueryFilterLimit
	}

	logger = logger.WithValue(keys.SearchQueryKey, query).WithValue(keys.FilterLimitKey, limit)

	// TODO: actually get search query in here
	params := &generated.SearchForValidVesselsParams{
		Q:     query,
		Limit: int(limit),
	}
	res, err := c.authedGeneratedClient.SearchForValidVessels(ctx, params, c.queryFilterCleaner)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "retrieving valid vessel")
	}
	defer c.closeResponseBody(ctx, res)

	var apiResponse types.APIResponse[[]*types.ValidVessel]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "retrieving valid vessels")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	return apiResponse.Data, nil
}

// GetValidVessels retrieves a list of valid vessels.
func (c *Client) GetValidVessels(ctx context.Context, filter *types.QueryFilter) (*types.QueryFilteredResult[types.ValidVessel], error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if filter == nil {
		filter = types.DefaultQueryFilter()
	}
	logger = filter.AttachToLogger(logger)
	tracing.AttachQueryFilterToSpan(span, filter)

	params := &generated.GetValidVesselsParams{}
	c.copyType(params, filter)

	res, err := c.authedGeneratedClient.GetValidVessels(ctx, params, c.queryFilterCleaner)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "retrieving valid vessels")
	}
	defer c.closeResponseBody(ctx, res)

	var apiResponse types.APIResponse[[]*types.ValidVessel]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "loading valid vessels list response")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	output := &types.QueryFilteredResult[types.ValidVessel]{
		Data:       apiResponse.Data,
		Pagination: *apiResponse.Pagination,
	}

	return output, nil
}

// CreateValidVessel creates a valid vessel.
func (c *Client) CreateValidVessel(ctx context.Context, input *types.ValidVesselCreationRequestInput) (*types.ValidVessel, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if input == nil {
		return nil, ErrNilInputProvided
	}

	if err := input.ValidateWithContext(ctx); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "validating input")
	}

	body := generated.CreateValidVesselJSONRequestBody{}
	c.copyType(&body, input)

	res, err := c.authedGeneratedClient.CreateValidVessel(ctx, body)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "creating valid vessel")
	}
	defer c.closeResponseBody(ctx, res)

	var apiResponse types.APIResponse[*types.ValidVessel]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "loading valid vessel creation response")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	return apiResponse.Data, nil
}

// UpdateValidVessel updates a valid vessel.
func (c *Client) UpdateValidVessel(ctx context.Context, validVessel *types.ValidVessel) error {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if validVessel == nil {
		return ErrNilInputProvided
	}
	logger = logger.WithValue(keys.ValidVesselIDKey, validVessel.ID)
	tracing.AttachToSpan(span, keys.ValidVesselIDKey, validVessel.ID)

	body := generated.UpdateValidVesselJSONRequestBody{}
	c.copyType(&body, validVessel)

	res, err := c.authedGeneratedClient.UpdateValidVessel(ctx, validVessel.ID, body)
	if err != nil {
		return observability.PrepareAndLogError(err, logger, span, "updating valid vessel")
	}
	defer c.closeResponseBody(ctx, res)

	var apiResponse types.APIResponse[*types.ValidVessel]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return observability.PrepareAndLogError(err, logger, span, "loading valid vessel update response")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return err
	}

	return nil
}

// ArchiveValidVessel archives a valid vessel.
func (c *Client) ArchiveValidVessel(ctx context.Context, validVesselID string) error {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if validVesselID == "" {
		return ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.ValidVesselIDKey, validVesselID)
	tracing.AttachToSpan(span, keys.ValidVesselIDKey, validVesselID)

	res, err := c.authedGeneratedClient.ArchiveValidVessel(ctx, validVesselID)
	if err != nil {
		return observability.PrepareAndLogError(err, logger, span, "archiving valid vessel")
	}
	defer c.closeResponseBody(ctx, res)

	var apiResponse types.APIResponse[*types.ValidVessel]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return observability.PrepareAndLogError(err, logger, span, "loading valid vessel archive response")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return err
	}

	return nil
}
